### ISSUE TITLE
Suggestion about adding new protocol bindings/consumers

### ISSUE THREAD

**Issue by erkann-sen** [eclipse/thingweb.node-wot#267]
Hi everyone,
I am currently writing my master thesis, which is about interoperability of protocols/systems with wot.
Till now, I made [Mozilla webthings](https://github.com/erkann-sen/wot-mozilla) and [openHAB](https://github.com/erkann-sen/wot-openhab/) consumable with node-wot with different methods.
What I wish is to somehow bring this functionality to node-wot library. 
But I know that, adding a binding or helper for every platform makes the main library fatter and harder to maintain and also creates an illusion that for every platform there is another binding needed, which makes node-wot library seem hard to use and there is a discussion about what to make with Oracle and Fujitsu bindings.

When I am working with openHAB, I found their solution to this system, can be also useful for node-wot.
They have two different but connecting repos. A core repo, which has the main functionality of the system and ana add-ons repo which has several bindings or helpers for other systems.

I suggest we should have a main node-wot library, which has only the main bindings like MQTT, HTTP etc.
In another repo, we can have vendor specific bindings and helpers, all has their own readme file, which explains how to use. In this way it is also encouraged to write people more bindings to make other protocols work easily.

Advantages that I can think of are:

1. Easily readable and less complex main code
2. Easily maintainable main repo.
3. Adding another bindings are easier because it does not mess up with the main repo, which can increase the adoption of the system 
4. The discussion about what to do with vendor specific bindings will end

**Comment by relu91** [comment]
I totally support this idea. I have been tinkering for a while about the possibility to move node-wot to a full plug-in based architecture. In my mind, not only protocol bindings should be provided as add-ons but also security schemas implementations, content-encoders/decoders, and (maybe) system APIs. As you mentioned there are quite a few benefits but also, I want to add, that this pattern is well-known in the node/JS community. Most of the big libraries are plug-in based like Vue.js, Jquery, Rollup... (just to mention a few). My personal opinion is that this open plugin architecture contributed a lot to the wide success of this open-source JS libraries.  

Furthermore, this open plug-in architecture will cope with the inherent complexity of Thing Descriptors. As @mlagally mentioned during different calls, it might even be impossible to build a client that is able to correctly understand ALL the possible TDs. That's why WoT Profiles are been explored and if they are adopted we could even tag different plug-ins to be compliant to one or more WoT profile.  

As your suggesting the first step could be to create a dedicated repo for community curated protocol bindings plugins but to fully achieve this goal (i.e. providing extension points for security and content-encoders), we have to rethink some parts of node-wot. Anyway, starting from just protocol bindings might be enough to explore the solution and later extend it to the other points. 

So long story short, let's go with it from my side 👍

**Comment by egekorkan** [comment]
Related to #236

**Comment by sebastiankb** [comment]
maybe a good orientation can be https://homebridge.io/ and its way to find plugins

**Comment by danielpeintner** [comment]
I think in the ideal and final situation I agree with all that has been said before.

However, at the moment the scripting API and some other aspects are probably not full *stable* yet.  Assuming we have *breaking* changes it is easier to maintain code in a single repo.

Having said that, I totally support creating different repos in the *near* future but my current take on this is that it is still to early....

**Comment by danielpeintner** [comment]
Summary of discussion between @egekorkan @wiresio  and @danielpeintner 

* the current `packages/xyz` structure with independent projects is already a good step foward
* we still need to find a good solution for ContentSerdes since every package might have its own codec and should not be part of (or blow up) `core`
* we also need to look into Eclipse licensing issues and "where" these new repos should be created, e.g., in Eclipse space, elsewhere, ...?
* similar we need to decide (or propose) best practices how the packages are named, see https://github.com/eclipse/thingweb.node-wot/issues/236

**Comment by danielpeintner** [comment]
>     * we still need to find a good solution for ContentSerdes since every package might have its own codec and should not be part of (or blow up) `core`

I believe w.r.t. ContentSerdes  PR https://github.com/eclipse/thingweb.node-wot/pull/304 should provide a solution/example.
